### PR TITLE
Implement pem_read_key directly through OSSL_DECODER

### LINKED ISSUE #14788: Calling PEM_read_bio_PrivateKey() and PEM_read_bio_X509() on the same BIO is broken
If there is both private key and certificate in a file and it is read by calling  PEM_read_bio_PrivateKey() and then PEM_read_bio_X509(), the X509 decoding will fail with the current master because when decoding the key the file is completely read and the X509 decoding does not have any data to read.

### PR DESCRIPTION
**Pull request by t8m**
Using OSSL_STORE is too heavy and breaks things.

There were also needed various fixes mainly for missing proper
handling of the SM2 keys in the OSSL_DECODER.

Fixes #14788

##### Checklist

- [x] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
This is still WIP because there is missing legacy PEM handling.

**Comment by t8m** [comment]
> This is still WIP because there is missing legacy PEM handling.

As reflected in the external GOST engine test failure.

**Comment by beldmit** [comment]
> As reflected in the external GOST engine test failure.

Yes :)

File: test/sslapitest.c [hunk 1]
```diff
@@ -8080,12 +8089,13 @@ static int test_sigalgs_available(int idx)
             goto end;
     }
 
-    if (!TEST_int_eq(SSL_CTX_use_certificate_file(sctx, cert2,
+    if (idx != 5
+        && (!TEST_int_eq(SSL_CTX_use_certificate_file(sctx, cert2,
                                                   SSL_FILETYPE_PEM), 1)
```
Comment (levitte) [inline test/sslapitest.c:8094]: Nit: indentation

**Comment by t8m** [comment]
Ready for review now.

**Comment by t8m** [comment]
I do not think the non-caching build failure is relevant. This seems to be some intermittent leak.

**Comment by t8m** [comment]
Ping for reviews

**Comment by t8m** [comment]
Rebased again, ping again for reviews!

File: crypto/pem/pem_pkey.c [hunk 2]
```diff
@@ -20,82 +20,198 @@
...
+            goto p8err;
+        ret = evp_pkcs82pkey_legacy(p8inf, libctx, propq);
+        if (x != NULL) {
+            EVP_PKEY_free(*x);
+            *x = ret;
+        }
+        PKCS8_PRIV_KEY_INFO_free(p8inf);
+    } else if ((slen = ossl_pem_check_suffix(nm, "PRIVATE KEY")) > 0) {
+        const EVP_PKEY_ASN1_METHOD *ameth;
+        ameth = EVP_PKEY_asn1_find_str(NULL, nm, slen);
```
Comment (paulidale) [inline crypto/pem/pem_pkey.c:150]: blank line

**Review by paulidale — APPROVED** [review]
One formatting NIT.

**Comment by t8m** [comment]
Rebased to resolve the conflict in checksums

**Comment by t8m** [comment]
One more clean rebase to test the FIPS checksums action

**Comment by t8m** [comment]
And one more

**Comment by t8m** [comment]
Hooray! The labeling works.

**Comment by t8m** [comment]
Merged to master. Thank you for the review!

**Pull request closed without merging**